File: mediamon/mediamon.py
import logging
import prometheus_client
from mediamon.version import version
from mediamon.configuration import print_configuration
from mediamon.transmission import TransmissionProbe
from mediamon.xxxarr import MonitorProbe
from mediamon.plex import PlexServer
from pimetrics.scheduler import Scheduler


def initialise(config):
    scheduler = Scheduler()

    if 'transmission' in config.services:
        try:
            scheduler.register(
                TransmissionProbe(config.services['transmission']['host']),
                config.services['transmission']['interval']
            )
        except KeyError as e:
            logging.warning(f'transmission config missing {e}. Skipping')

    if 'sonarr' in config.services:
        try:
            scheduler.register(
                MonitorProbe(
                    config.services['sonarr']['host'],
                    MonitorProbe.App.sonarr,
                    config.services['sonarr']['apikey']
                ),
                config.services['sonarr']['interval']
            )
        except KeyError as e:
            logging.warning(f'sonarr config missing {e}. Skipping')

    if 'radarr' in config.services:
        try:
            scheduler.register(
                MonitorProbe(
                    config.services['radarr']['host'],
                    MonitorProbe.App.radarr,
                    config.services['radarr']['apikey']),
                config.services['radarr']['interval']
            )
        except KeyError as e:
            logging.warning(f'radarr config missing {e}. Skipping')

    if 'plex' in config.services:
        try:
            scheduler.register(
                PlexServer(config.services['plex']['username'], config.services['plex']['password']),
                config.services['plex']['interval']
            )
        except KeyError as e:
            logging.warning(f'plex config missing {e}. Skipping')

    if len(scheduler.scheduled_items) == 0:
        logging.error('No services defined')

    return scheduler


def mediamon(config):
    logging.basicConfig(format='%(asctime)s - %(levelname)s - %(message)s', datefmt='%Y-%m-%d %H:%M:%S',
                        level=logging.DEBUG if config.debug else logging.INFO)
    logging.info(f'Starting mediamon v{version}')
    logging.info(f'Configuration: {print_configuration(config)}')

    prometheus_client.start_http_server(config.port)

    scheduler = initialise(config)
    if config.once:
        scheduler.run(once=True)
    else:
        while True:
            scheduler.run(duration=config.interval)
    return 0

File: mediamon.go
package main

import (
	"github.com/clambin/gotools/metrics"
	"github.com/clambin/mediamon/collectors/bandwidth"
	"github.com/clambin/mediamon/collectors/connectivity"
	"github.com/clambin/mediamon/collectors/plex"
	"github.com/clambin/mediamon/collectors/transmission"
	"github.com/clambin/mediamon/collectors/xxxarr"
	"github.com/clambin/mediamon/services"
	"github.com/clambin/mediamon/version"
	"github.com/prometheus/client_golang/prometheus"
	log "github.com/sirupsen/logrus"
	"gopkg.in/alecthomas/kingpin.v2"
	"net/http"
	"os"
	"path/filepath"
)

type configuration struct {
	Port     int
	Debug    bool
	Services *services.Config
}

func main() {
	var servicesFilename string

	cfg := configuration{}

	a := kingpin.New(filepath.Base(os.Args[0]), "media monitor")

	a.Version(version.BuildVersion)
	a.HelpFlag.Short('h')
	a.VersionFlag.Short('v')
	a.Flag("debug", "Log debug messages").BoolVar(&cfg.Debug)
	a.Flag("port", "API listener port").Default("8080").IntVar(&cfg.Port)
	a.Flag("file", "Service configuration file").Required().StringVar(&servicesFilename)

	_, err := a.Parse(os.Args[1:])
	if err != nil {
		a.Usage(os.Args[1:])
		os.Exit(2)
	}

	if cfg.Debug {
		log.SetLevel(log.DebugLevel)
	}

	if cfg.Services, err = services.ParseConfigFile(servicesFilename); err != nil {
		log.WithFields(log.Fields{"err": err, "filename": servicesFilename}).Warning("unable to parse Services file")
		os.Exit(3)
	}

	log.WithField("version", version.BuildVersion).Info("media monitor starting")

	// Transmission Collector
	if cfg.Services.Transmission.URL != "" {
		log.WithField("url", cfg.Services.Transmission.URL).Info("monitoring Transmission")
		prometheus.DefaultRegisterer.MustRegister(transmission.NewCollector(
			cfg.Services.Transmission.URL,
			cfg.Services.Transmission.Interval,
		))
	}

	// Sonarr Collector
	if cfg.Services.Sonarr.URL != "" {
		log.WithField("url", cfg.Services.Sonarr.URL).Info("monitoring Sonarr")
		prometheus.DefaultRegisterer.MustRegister(xxxarr.NewCollector(
			cfg.Services.Sonarr.URL,
			cfg.Services.Sonarr.APIKey,
			"sonarr",
			cfg.Services.Sonarr.Interval,
		))
	}

	// Radarr Collector
	if cfg.Services.Radarr.URL != "" {
		log.WithField("url", cfg.Services.Radarr.URL).Info("monitoring Radarr")
		prometheus.DefaultRegisterer.MustRegister(xxxarr.NewCollector(
			cfg.Services.Radarr.URL,
			cfg.Services.Radarr.APIKey,
			"radarr",
			cfg.Services.Radarr.Interval,
		))
	}

	// Plex Collector
	if cfg.Services.Plex.URL != "" {
		log.WithField("url", cfg.Services.Plex.URL).Info("monitoring Plex")
		prometheus.DefaultRegisterer.MustRegister(plex.NewCollector(
			cfg.Services.Plex.URL,
			cfg.Services.Plex.UserName,
			cfg.Services.Plex.Password,
			cfg.Services.Plex.Interval,
		))
	}

	// Bandwidth Probe
	if cfg.Services.OpenVPN.Bandwidth.FileName != "" {
		log.WithField("filename", cfg.Services.OpenVPN.Bandwidth.FileName).Info("monitoring OpenVPN Bandwidth usage")
		prometheus.DefaultRegisterer.MustRegister(bandwidth.NewCollector(
			cfg.Services.OpenVPN.Bandwidth.FileName,
			cfg.Services.OpenVPN.Bandwidth.Interval,
		))
	}

	// Connectivity Probe
	if cfg.Services.OpenVPN.Connectivity.ProxyURL != nil {
		log.WithField("proxyURL", cfg.Services.OpenVPN.Connectivity.ProxyURL).Info("monitoring OpenVPN connectivity")
		prometheus.DefaultRegisterer.MustRegister(connectivity.NewCollector(
			cfg.Services.OpenVPN.Connectivity.Token,
			cfg.Services.OpenVPN.Connectivity.Proxy,
			cfg.Services.OpenVPN.Connectivity.Interval,
		))
	}

	server := metrics.NewServer(cfg.Port)
	err = server.Run()
	if err != http.ErrServerClosed {
		log.WithField("err", err).Error("Failed to start Prometheus http handler")
	}

	log.Info("mediamon exiting")
}
